hoomd/md/pair/alch.py

```diff
@@ -0,0 +1,220 @@
+    def mass(self, M):
+        if self._attached:
+            self._cpp_obj.mass = M
+        else:
+            self._mass = M
+
+    @log
+    def value(self):
+        return self.force.params[self.typepair][self.name] * (
+            self._cpp_obj.alpha if self._attached else 1.)
+
+    @log(default=False, requires_run=True)
+    def avalue(self):
+        return self._cpp_obj.alpha
+
+    @log(default=False, requires_run=True)
+    def amomentum(self):
+        return self._cpp_obj.alpha
+
+    @log(default=False, requires_run=True)
+    def mu(self):
+        return self._cpp_obj.mu
+
+    @log(default=False, requires_run=True, category='particle')
+    def aforces(self):
+        return self._cpp_obj.forces
+
+    @log(requires_run=True)
+    def net_aforce(self):
+        return self._cpp_obj.net_force()
+
+    def _enable(self):
+        self.force._cpp_obj.enableAlchemicalPairParticle(self._cpp_obj)
+
+
+class _AlchemicalMethods(_HOOMDBaseObject):
+
+    def __init__(self):
+        self.alchemical_particles = self.AlchemicalParticleAccess(self)
+
+    class AlchemicalParticleAccess(TypeParameterDict):
+
+        def __init__(self, outer):
+            self.outer = outer
+            super().__init__(SetOnce(AlchemicalMDParticle), len_keys=2)
+
+        def _validate_and_split_alchem(self, key):
+            if isinstance(key, tuple) and len(key) == self._len_keys:
+                for param, typepair in zip(key, reversed(key)):
+                    if set(self.outer._alchemical_parameters).issuperset(
+                            self._validate_and_split_len_one(param)) and set(
+                                self.outer._simulation.state.particle_types
+                            ).issuperset(
+                                self._validate_and_split_len_one(typepair)):
+                        break
+                else:
+                    raise KeyError(
+                        key,
+                        'Not a valid combination of particle types and alchemical parameter keys.'
+                    )
+                return tuple(self._yield_keys(typepair)), tuple(
+                    self._validate_and_split_len_one(param))
+
+        def __getitem__(self, key):
+            typepair, param = self._validate_and_split_alchem(key)
+            vals = dict()
+            for t in typepair:
+                for p in param:
+                    k = (t, p)
+                    if not k in self._dict:
+                        self._dict[k] = AlchemicalMDParticle(self.outer, p, t)
+                    vals[k] = self._dict[k]
+            if len(vals) > 1:
+                return vals
+            else:
+                return vals[k]
+
+        def __setitem__(self, key, val):
+            raise NotImplementedError
+
+        def __contains__(self, key):
+            key = self._validate_and_split_alchem(key)
+            return key in self._dict.keys()
+
+        def keys(self):
+            return self._dict.keys()
```

Comment:
My somewhat convoluted attempt to make accessing/creating the AlchMDParticles of an alchemical pair potential easy.